Explorer/Assets/DCL/SDKComponents/SceneUI/Systems/UIPointerEvents/UIPointerEventsSystem.cs

```diff
@@ -0,0 +1,74 @@
+﻿using Arch.Core;
+using Arch.System;
+using Arch.SystemGroups;
+using Arch.SystemGroups.Throttling;
+using CRDT;
+using CrdtEcsBridge.ECSToCRDTWriter;
+using DCL.Diagnostics;
+using DCL.ECSComponents;
+using DCL.SDKComponents.SceneUI.Components;
+using DCL.SDKComponents.SceneUI.Systems.UITransform;
+using ECS.Abstract;
+using ECS.Groups;
+using SceneRunner.Scene;
+using UnityEngine.UIElements;
+
+namespace DCL.SDKComponents.SceneUI.Systems.UIPointerEvents
+{
+    [UpdateInGroup(typeof(SyncedSimulationSystemGroup))]
+    [UpdateAfter(typeof(UITransformUpdateSystem))]
+    [LogCategory(ReportCategory.SCENE_UI)]
+    [ThrottlingEnabled]
+    public partial class UIPointerEventsSystem : BaseUnityLoopSystem
+    {
+        private readonly ISceneStateProvider sceneStateProvider;
+        private readonly IECSToCRDTWriter ecsToCRDTWriter;
+
+        public UIPointerEventsSystem(World world, ISceneStateProvider sceneStateProvider, IECSToCRDTWriter ecsToCRDTWriter) : base(world)
+        {
+            this.sceneStateProvider = sceneStateProvider;
+            this.ecsToCRDTWriter = ecsToCRDTWriter;
+        }
+
+        protected override void Update(float _)
+        {
+            UpdatePointerEventsQuery(World);
+        }
+
+        [Query]
+        private void UpdatePointerEvents(ref PBPointerEvents sdkModel, ref UITransformComponent uiTransformComponent, CRDTEntity sdkEntity)
+        {
+            if (!sdkModel.IsDirty)
+                return;
+
+            foreach (var pEvent in sdkModel.PointerEvents)
+            {
+                switch (pEvent.EventType)
+                {
+                    case PointerEventType.PetDown:
+                        uiTransformComponent.Transform.RegisterPointerDownCallback(_ => AppendMessage(sdkEntity, pEvent.EventInfo.Button, PointerEventType.PetDown));
```

Comment:
You introduced a dangerous closure here: you capture `CRDTEntity` and `pEvent`. It results in:
- unnecessary allocations 
- disconnects from the component/entity life-cycle.

Event handling contradicts ECS lifecycle as it disconnects the moment of the callback from the system update order and its possible destruction.

It should be improved:
- I propose you to gather indices and `PointerEventType` being clicked in a separate structure (you may keep it in `UITransformComponent`), 
- Then in a regular query iterate over that data and call `AppendMessage` accordingly
- Thus you will be able to take all the required data from the query without closures

Reply:
As we discussed via slack, I've applied the corresponding changes in order to move the AppendMessage call outside of the callbacks 👍 

Refactor done here: https://github.com/decentraland/unity-explorer/pull/290/commits/5ba26fe5c09e4710962d2b7038a4a9989fea96ab